openretina/utils/plotting.py

```diff
@@ -388,3 +400,370 @@ def legend_without_duplicate_labels(ax=None) -> None:
+    """
+
+    # Get number of neurons and normalize activity
+    n_neurons = np.min(roi_mask) * -1
+    activity_normalized = np.clip(activity / max_activity, 0, 1)
+
+    # Pre-compute color mapping if cell types are provided
+    color_map = {}
+    if cell_types is not None:
+        max_type = np.max(cell_types) if max_cell_type is None else max_cell_type
+        min_type = np.min(cell_types) if min_cell_type is None else min_cell_type
+
+        # Create broad type ranges
+        if type_boundaries is None:
+            type_boundaries = []
+        type_boundaries = sorted([min_type] + type_boundaries + [max_type])
+
+        # For each broad type range, assign a base saturation
+        saturations = {}
+        for i in range(len(type_boundaries) - 1):
+            start, end = type_boundaries[i], type_boundaries[i + 1]
+            types_in_range = np.where((cell_types >= start) & (cell_types <= end))[0]
+            base_saturation = 0.6 + (i * (0.3 / len(type_boundaries)))  # Vary saturation slightly for each broad type
+            for type_idx in types_in_range:
+                saturations[cell_types[type_idx]] = base_saturation
+
+        # Assign hues based on cell type numbers
+        for cell_type in np.unique(cell_types):
+            # Map cell type to hue (0-1)
+            hue = (cell_type - min_type) / (max_type - min_type)
+            saturation = saturations[cell_type]
+            color_map[cell_type] = (hue, saturation)
+
+    # Pre-compute centroids for each ROI if we're visualizing IDs
+    centroids = {}
+    if visualize_ids:
+        for i in range(n_neurons):
+            if roi_ids is not None and (i + 1) not in roi_ids:
+                continue
+            neuron_mask = roi_mask == -(i + 1)
+            if np.any(neuron_mask):
+                centroids[i] = center_of_mass(neuron_mask)
+
+    def render_frame(frame_idx):
+        """Render a single frame using matplotlib"""
+        # Create figure and axis
+        fig, ax = plt.subplots(figsize=figsize)
+        plt.close()
+
+        # Create frame image (now in RGB)
+        frame_image = np.zeros((*roi_mask.shape, 3))
+
+        # Update each neuron's color based on type and brightness based on activity
+        if roi_ids is not None:
+            for i, roi_id in enumerate(roi_ids):
+                neuron_mask = roi_mask == -roi_id
+                activity_val = activity_normalized[i, frame_idx]
+
+                if cell_types is not None:
+                    # Get pre-computed hue and saturation for this cell type
+                    hue, saturation = color_map[cell_types[i]]
+                    # Convert HSV to RGB
+                    rgb = mcolors.hsv_to_rgb([hue, saturation, activity_val])
+                    frame_image[neuron_mask] = rgb
+                else:
+                    # If no cell types provided, use grayscale
+                    frame_image[neuron_mask] = [activity_val] * 3
```

Comment:
```suggestion
                    frame_image[neuron_mask] = activity_val * 3
```
Does this not work here?

Reply:
It needs to duplicate the frame value to three channels